cpu/atmega2560/periph/gpio.c

```diff
@@ -53,23 +195,262 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
 
 int gpio_read(gpio_t dev)
 {
-    return -1;
+    uint8_t pin;
+
+    if (dev > GPIO_NUMOF)
+    {
+        return -1;
+    }
+
+    pin = gpio_pin_map[dev];
+
+    switch (dev)
+    {
+        case GPIO_0:
+            if ( GPIO_0_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_1:
+            if ( GPIO_1_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_2:
+            if ( GPIO_2_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_3:
+            if ( GPIO_3_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_4:
+            if ( GPIO_4_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_5:
+            if ( GPIO_5_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_6:
+            if ( GPIO_6_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_7:
+            if ( GPIO_7_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_8:
+            if ( GPIO_8_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_9:
+            if ( GPIO_9_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_10:
+            if ( GPIO_10_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
+        case GPIO_11:
+            if ( GPIO_11_PORT & (1 << pin))
+            {
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+            break;
```

Comment:
Yes, i would like to implement it this way, but i don't know how, because the PORTx seems not to be constant and the initializer element in the port map has to be.
hmm.. i tried it already before..